lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -247,58 +247,59 @@ void LocalEnforcer::sync_sessions_on_restart(std::time_t current_time) {
                    << " during record aggregation";
       continue;
     }
-    sessions_with_active_flows.insert(record.sid());
+    auto& session          = **session_it;
+    const auto& session_id = session->get_session_id();
     if (record.bytes_tx() > 0 || record.bytes_rx() > 0) {
-      MLOG(MINFO) << record.sid() << " used " << record.bytes_tx()
+      MLOG(MINFO) << session_id << " used " << record.bytes_tx()
                   << " tx bytes and " << record.bytes_rx()
                   << " rx bytes for rule " << record.rule_id();
     }
-    // Update sessions
-    for (const auto& session : it->second) {
-      SessionStateUpdateCriteria& uc =
-          session_update[record.sid()][session->get_session_id()];
-      session->add_rule_usage(
-          record.rule_id(), record.bytes_tx(), record.bytes_rx(), uc);
-    }
+    SessionStateUpdateCriteria& uc = session_update[imsi][session_id];
+    session->add_rule_usage(
+        record.rule_id(), record.bytes_tx(), record.bytes_rx(), uc);
+    sessions_with_active_flows.insert(ImsiAndSid(imsi, session_id));
   }
   complete_termination_for_released_sessions(
       session_map, sessions_with_active_flows, session_update);
 }
 
 void LocalEnforcer::complete_termination_for_released_sessions(
-    SessionMap& session_map, std::unordered_set<std::string> sessions_with_active_flows,
+    SessionMap& session_map,
+    std::unordered_set<ImsiAndSid> sessions_with_active_flows,
     SessionUpdate& session_update) {
   // Iterate through sessions and notify that report has finished. Terminate any
   // sessions that can be terminated.
-  std::vector<std::pair<std::string, std::string>> imsi_to_terminate;
+  std::vector<ImsiAndSid> imsi_to_terminate;
   for (const auto& session_pair : session_map) {
     const std::string imsi = session_pair.first;
     for (const auto& session : session_pair.second) {
       const std::string session_id = session->get_session_id();
       // If we did not receive a rule record for the session, this means
       // PipelineD has reported all usage for the session
       if (session->get_state() == SESSION_RELEASED &&
-          sessions_with_active_flows.find(imsi) == sessions_with_active_flows.end()) {
+          sessions_with_active_flows.find(ImsiAndSid(imsi, session_id)) ==
+              sessions_with_active_flows.end()) {
         imsi_to_terminate.push_back(std::make_pair(imsi, session_id));
       }
     }
   }
+  // Do the actual termination in a separate loop since this can modify the
+  // session map structure
   for (const auto& imsi_sid_pair : imsi_to_terminate) {
-    auto imsi                      = imsi_sid_pair.first;
-    auto session_id                = imsi_sid_pair.second;
-    SessionStateUpdateCriteria& uc = session_update[imsi][session_id];
-    complete_termination(session_map, imsi, session_id, uc);
+    complete_termination(
+        session_map, imsi_sid_pair.first, imsi_sid_pair.second, session_update);
```

Comment:
I rather prefer this here

```
    auto imsi                      = imsi_sid_pair.first;	   
    auto session_id                = imsi_sid_pair.second;
    complete_termination(session_map, imsi, session_id, session_update);
```
But just a preference


Also, I would suggest not to use `sid`. I have seen that in other places in reference to subscriberId. So i would use `imsi_sessionid_pair`. Adds more characters, but saves time to the reader trying to figure out if this is session or subscriber

Reply:
Yeah, good call. I will fix that.